content/algorithms/planarity/planarity.md

```diff
@@ -0,0 +1,939 @@
+
+---
+
+
+
+***What is  Left-Right Planarity Test***
+
+The Left-Right Planarity Test is a linear-time algorithm used to check whether a graph is planar—that is, drawable in the plane without edge crossings. This test forms the core of the networkx.check_planarity() function.
+
+***How It Works (Conceptually):***
+
+
+
+1. *Depth-First Search (DFS):*  
+The algorithm starts by building a DFS tree of the input graph, which helps classify edges into tree edges and back edges.
+2. *Lowpoints and Embedding Constraints:*  
+For each vertex, the algorithm computes "lowpoints"—the smallest reachable ancestor in DFS. It uses this to decide how back edges can be placed to the left or right of tree edges without crossings.
+3. *Conflict Detection:*  
+The algorithm tracks embedding constraints between edges. If it detects a conflict (e.g., an edge that can't be placed without violating planarity), it concludes the graph is non-planar.
+4. *Constructs a Planar Embedding or Finds Obstruction:*  
+If the graph is planar, it returns a combinatorial embedding (rotation system). If not, it can identify a Kuratowski subgraph (e.g., K₅ or K₃,₃).
+
+
+
+
+
+
+```python
+from functools import singledispatch
+```
+
+
+```python
+from collections import defaultdict
+import networkx as nx
+import numpy as np
+import matplotlib.pyplot as plt
+```
+
+
+```python
+__all__ = ["check_planarity", "is_planar", "PlanarEmbedding"]
+```
+
+## Parameters
+
+G: NetworkX graph
+counterexample : bool
+A Kuratowski subgraph (to prove non planarity) is only returned if set to true.
+
+
+What is ***Kuratowski subgraph*** ?
+
+A Kuratowski subgraph is a specific subgraph that serves as evidence that a graph is not planar. According to Kuratowski’s Theorem, a graph is non-planar if and only if it contains a subgraph that is a subdivision (i.e., homeomorphic copy) of either **K₅ (the complete graph on 5 vertices)** or **K₃,₃ (the complete bipartite graph with partitions of size 3)**. These subgraphs are called Kuratowski subgraphs.
+
+They may not appear in their exact form but can be present in a stretched-out version where edges are replaced by paths. NetworkX uses this principle internally: *if check_planarity() returns False, it indicates the existence of a Kuratowski subgraph obstructing planar embedding.*
+
+---
+
+
+
+## Returns
+
+The function returns a tuple of the form:
+
+
+*   is_planar (bool):
+  *   True if the graph is planar.
+  *   False if the graph is non-planar.
+
+*   certificate (NetworkX graph):
+  *   If is_planar is True, certificate will be a PlanarEmbedding object representing a valid planar embedding of the graph.
+  *   If is_planar is False, certificate will be a Kuratowski subgraph, which is a subgraph of the original graph that proves the graph is non-planar by being a homeomorphic copy of K₅ or K₃,₃.
+
+---
+
+
+
+
+```python
+from networkx.algorithms.planarity import LRPlanarity
+```
+
+
+```python
+def is_planar(G):
```

Comment:
Are these functions part of NetworkX?  If so, you don’t need to define them here. You can call them from e.g. `nx.is_planar`